tachyon/crypto/sumcheck/sumcheck_verifier.h

```diff
@@ -0,0 +1,267 @@
+// Copyright 2022 arkworks contributors
+// Use of this source code is governed by a MIT/Apache-2.0 style license that
+// can be found in the LICENSE-MIT.arkworks and the LICENCE-APACHE.arkworks
+// file.
+
+#ifndef TACHYON_CRYPTO_SUMCHECK_SUMCHECK_VERIFIER_H_
+#define TACHYON_CRYPTO_SUMCHECK_SUMCHECK_VERIFIER_H_
+
+#include <utility>
+#include <vector>
+
+#include "tachyon/crypto/sumcheck/sumcheck_prover_msg.h"
+#include "tachyon/crypto/sumcheck/sumcheck_verifier_msg.h"
+#include "tachyon/crypto/sumcheck/sumcheck_verifying_key.h"
+
+namespace tachyon::crypto {
+
+template <typename F>
+F InterpolateUniPoly(const std::vector<F>& poly, const F& evaluation_point);
+
+// Subclaim created when verifier is convinced.
+template <typename F>
+struct Subclaim {
+  // The multi-dimensional point this multilinear evaluation is evaluated at.
+  std::vector<F> point;
+  // The expected evaluation.
+  F expected_evaluation;
+};
+
+template <typename MLE>
+class SumcheckVerifier {
+ public:
+  using F = typename MLE::Field;
+
+  SumcheckVerifier() = default;
+  explicit SumcheckVerifier(const SumcheckVerifyingKey& key)
+      : num_variables_(key.num_variables),
+        max_evaluations_(key.max_evaluations) {
+    randomness_.reserve(key.num_variables);
+    polynomials_received_.reserve(key.num_variables);
+  }
+  explicit SumcheckVerifier(SumcheckVerifyingKey&& key)
+      : num_variables_(key.num_variables),
+        max_evaluations_(key.max_evaluations) {
+    randomness_.reserve(key.num_variables);
+    polynomials_received_.reserve(key.num_variables);
+  }
+
+  // Runs a "round" given a |SumcheckProverMsg|, sampling and storing a random
+  // value.
+  //
+  // While a true verifier round should perform actual verification, |Round()|
```

Comment:
What do you mean by a "true verifier"?

Reply:
in terms of the general protocol, the rounds should go like this
prover creates proof ->
<- verifier verifies proof & sends challenge back (aka a true verifier round)
prover uses challenge to create proof ->
etc...

yet, the prover and verifier currently implemented work like so:
prover creates proof -> 
<- verifier sends challenge & stores proof (technially not a true verifier round)
prover uses challenge to create proof -> 
...
after all these rounds are finished:
<- verifier verifies all proofs at one time